File: src/SimpleLevelEditor.Formats/Types/EntityConfig/EntityShapeDescriptor.fs
namespace SimpleLevelEditor.Formats.Types.EntityConfig

open SimpleLevelEditor.Formats.Types
open SimpleLevelEditor.Formats.Types.Level

type EntityShapeDescriptor =
    | Point of Visualization: PointEntityVisualization
    | Sphere of Color: Rgb
    | Aabb of Color: Rgb

    member this.GetDefaultEntityShape() =
        match this with
        | Point _  -> EntityShape.Point
        | Sphere _ -> EntityShape.Sphere 2f
        | Aabb _   -> EntityShape.Aabb System.Numerics.Vector3.One

    member this.GetTypeId() =
        match this with
        | Point _  -> nameof Point
        | Sphere _ -> nameof Sphere
        | Aabb _   -> nameof Aabb

    member this.DeepCopy() =
        match this with
        | Point visualization -> Point (visualization.DeepCopy())
        | Sphere color        -> Sphere color
        | Aabb color          -> Aabb color

File: src/SimpleLevelEditor.Formats.EntityConfig/EntityShapeDescriptor.cs
using Dunet;
using SimpleLevelEditor.Formats.Core;

namespace SimpleLevelEditor.Formats.EntityConfig;

[Union]
public partial record EntityShapeDescriptor
{
	public sealed partial record Point(PointEntityVisualization Visualization);
	public sealed partial record Sphere(Rgb Color);
	public sealed partial record Aabb(Rgb Color);

	// public EntityShape GetDefaultEntityShape()
	// {
	// 	return this switch
	// 	{
	// 		Point _ => new Point(new PointEntityVisualization.SimpleSphere(new Rgb(255, 255, 255), 1)),
	// 		Sphere _ => new Sphere(new Rgb(255, 255, 255)),
	// 		Aabb _ => new Aabb(new Rgb(255, 255, 255)),
	// 		_ => throw new InvalidOperationException()
	// 	};
	// }

	public string GetTypeId()
	{
		return this switch
		{
			Point => nameof(Point),
			Sphere => nameof(Sphere),
			Aabb => nameof(Aabb),
			_ => throw new InvalidOperationException(),
		};
	}

	public EntityShapeDescriptor DeepCopy()
	{
		return this switch
		{
			Point point => new Point(point.Visualization.DeepCopy()),
			Sphere sphere => new Sphere(sphere.Color),
			Aabb aabb => new Aabb(aabb.Color),
			_ => throw new InvalidOperationException(),
		};
	}
}
